File: libpandadna/SuitLegList.py
from SuitLeg import *
from DNASuitPoint import DNASuitPoint

class SuitLegList:
    def __init__(self, path, dnaStore, suitWalkSpeed, fromSky, toSky,
            fromSuitBuilding, toSuitBuilding, toToonBuilding):
        self.path = path
        self.dnaStore = dnaStore
        self.suitWalkSpeed = suitWalkSpeed
        self.fromSky = fromSky
        self.toSky = toSky
        self.fromSuitBuilding = fromSuitBuilding
        self.toSuitBuilding = toSuitBuilding
        self.toToonBuilding = toToonBuilding

        self.legs = []

        startPoint = self.path.getPoint(0)
        headingPoint = self.path.getPoint(1)
        zoneId = self.dnaStore.getSuitEdgeZone(startPoint.getIndex(),
                                               headingPoint.getIndex())
        startLeg = SuitLeg(self.getStartTime(0), zoneId, -1, startPoint,
                           headingPoint, self.getFirstLegType())
        self.legs.append(startLeg)

        for i in xrange(1, self.path.getNumPoints()):
            pointA = self.path.getPoint(i - 1)
            pointB = self.path.getPoint(i)
            zoneId = self.dnaStore.getSuitEdgeZone(pointA.getIndex(),
                                                   pointB.getIndex())
            landmarkBuildingIndex = pointA.getLandmarkBuildingIndex()
            
            if landmarkBuildingIndex == -1:
                landmarkBuildingIndex = pointB.getLandmarkBuildingIndex()
                
            leg = SuitLeg(self.getStartTime(i), zoneId, landmarkBuildingIndex,
                          pointA, pointB, self.getNextLegType(i - 1))
            self.legs.append(leg)

        endIndex = self.path.getNumPoints() - 1
        endPoint = self.path.getPoint(endIndex)
        endEdge = self.dnaStore.suitEdges[endPoint.getIndex()][0]
        zoneId = endEdge.getZoneId()
        endLeg = SuitLeg(self.getStartTime(endIndex), zoneId, -1, endPoint,
                         endPoint, self.getLastLegType())
        self.legs.append(endLeg)

    def getSuitLegType(self, pointTypeA, pointTypeB):
        if pointTypeA == DNASuitPoint.pointTypeMap['STREET_POINT']:
        
            if pointTypeB == DNASuitPoint.pointTypeMap['STREET_POINT']:
                return SuitLeg.TWalk
                
            elif pointTypeB == DNASuitPoint.pointTypeMap['FRONT_DOOR_POINT']:
                return SuitLeg.TToToonBuilding
                
            elif pointTypeB == DNASuitPoint.pointTypeMap['SIDE_DOOR_POINT']:
                return SuitLeg.TToSuitBuilding
                
        elif pointTypeA == DNASuitPoint.pointTypeMap['SIDE_DOOR_POINT']:
            return SuitLeg.TFromSuitBuilding
        
        elif pointTypeA == DNASuitPoint.pointTypeMap['COGHQ_IN_POINT']:
            return SuitLeg.TToCogHQ
        
        elif pointTypeA == DNASuitPoint.pointTypeMap['COGHQ_OUT_POINT']:
            return SuitLeg.TFromCogHQ
        
        return SuitLeg.TWalk

    def getNumLegs(self):
        return len(self.legs)

    def getLeg(self, index):
        return self.legs[index]

    def getType(self, index):
        return self.legs[index].getType()

    def getLegTime(self, index):
        return self.legs[index].getLegTime()

    def getZoneId(self, index):
        return self.legs[index].getZoneId()

    def getBlockNumber(self, index):
        return self.legs[index].getBlockNumber()

    def getPointA(self, index):
        return self.legs[index].getPointA()

    def getPointB(self, index):
        return self.legs[index].getPointB()

    def getStartTime(self, index):
        if index < self.getNumLegs():
            return self.legs[index].getStartTime()
            
        startTime = 0.0
        
        for legIndex in xrange(self.getNumLegs()):
            if legIndex == index:
                break
            startTime += self.getLegTime(legIndex)
            
        return startTime

    def getLegIndexAtTime(self, time, startLeg):
        endTime = 0.0
        
        for legIndex in xrange(self.getNumLegs()):
            endTime += self.getLegTime(legIndex)
            
            if endTime > time:
                break
                
        return legIndex

    def isPointInRange(self, point, lowTime, highTime):
        pointIndex = point.getIndex()
        startLegIndex = self.getLegIndexAtTime(lowTime, 0)
        endLegIndex = self.getLegIndexAtTime(highTime, pointIndex)
        
        for leg in self.legs[startLegIndex:endLegIndex + 1]:
            if leg.getPointA().getIndex() == pointIndex:
                return 1
            
            elif leg.getPointB().getIndex() == pointIndex:
                return 1
                
        return 0

    def getFirstLegType(self):
        pointTypeA = self.path.getPoint(0).getPointType()
        
        if pointTypeA == DNASuitPoint.pointTypeMap['STREET_POINT']:
            return SuitLeg.TFromSky
            
        elif pointTypeA == DNASuitPoint.pointTypeMap['COGHQ_OUT_POINT']:
            return SuitLeg.TFromCogHQ
            
        pointTypeB = self.path.getPoint(1).getPointType()
        
        return self.getSuitLegType(pointTypeA, pointTypeB)

    def getNextLegType(self, index):
        pointTypeA = self.path.getPoint(index).getPointType()
        pointTypeB = self.path.getPoint(index + 1).getPointType()
        return self.getSuitLegType(pointTypeA, pointTypeB)

    def getLastLegType(self):
        return SuitLeg.TToSky

    def __getitem__(self, key):
        return self.legs[key]
        

File: src/suit/SuitLegList.cxx
#include "SuitLegList.h"

SuitLegList::SuitLegList(PT(DNASuitPath) path, DNAStorage* store, double suit_walk_speed): m_path(path),
                         m_store(store), m_suit_walk_speed(suit_walk_speed)
{
    // Create first leg
    PT(DNASuitPoint) start_point = path->get_point(0);
    PT(DNASuitPoint) heading_point = path->get_point(1);
    add_leg(start_point, heading_point,
            start_point->get_point_type() == DNASuitPoint::SIDE_DOOR_POINT
            ? SuitLeg::T_from_suit_building : SuitLeg::T_from_sky, -1);

    // Create mid legs
    for (int i = 1; i < path->get_num_points(); i++)
    {
        PT(DNASuitPoint) point_a = path->get_point(i - 1);
        PT(DNASuitPoint) point_b = path->get_point(i);
        block_number_t landmark_building_index = point_a->get_landmark_building_index();
        if (landmark_building_index == -1)
            landmark_building_index = point_b->get_landmark_building_index();

        add_leg(point_a, point_b, get_next_leg_type(i - 1), landmark_building_index);
    }

    // Create last leg
    PT(DNASuitPoint) end_point;

    int end_index = path->get_num_points() - 1;
    start_point = path->get_point(end_index - 1);
    end_point = path->get_point(end_index);

    SuitLeg::LegType last_leg_type;
    if (end_point->get_point_type() == DNASuitPoint::FRONT_DOOR_POINT)
        last_leg_type = SuitLeg::T_to_toon_building;

    else if (end_point->get_point_type() == DNASuitPoint::SIDE_DOOR_POINT)
        last_leg_type = SuitLeg::T_to_suit_building;

    else
        last_leg_type = SuitLeg::T_to_sky;

    add_leg(start_point, end_point, last_leg_type,
            end_point->get_landmark_building_index());

    add_leg(start_point, end_point, SuitLeg::T_off, -1);
}

SuitLegList::~SuitLegList()
{
    for (leg_vec_t::iterator it = m_legs.begin(); it != m_legs.end(); it++)
        delete *it;
}

SuitLeg::LegType SuitLegList::get_suit_leg_type(DNASuitPoint::PointType a,
                                                DNASuitPoint::PointType b)
{
    if (a == DNASuitPoint::SIDE_DOOR_POINT)
        return SuitLeg::T_walk_to_street;

    if (b == DNASuitPoint::FRONT_DOOR_POINT || b == DNASuitPoint::SIDE_DOOR_POINT)
        return SuitLeg::T_walk_from_street;

    return SuitLeg::T_walk;
}

int SuitLegList::get_num_legs()
{
    return m_legs.size();
}

SuitLeg* SuitLegList::get_leg(int index)
{
    nassertr(index < get_num_legs(), nullptr);
    return m_legs.at(index);
}

SuitLeg::LegType SuitLegList::get_type(int index)
{
    return get_leg(index)->get_type();
}

double SuitLegList::get_leg_time(int index)
{
    return get_leg(index)->get_leg_time();
}

zone_id_t SuitLegList::get_zone_id(int index)
{
    return get_leg(index)->get_zone_id();
}

block_number_t SuitLegList::get_block_number(int index)
{
    return get_leg(index)->get_block_number();
}

PT(DNASuitPoint) SuitLegList::get_point_a(int index)
{
    return get_leg(index)->get_point_a();
}

PT(DNASuitPoint) SuitLegList::get_point_b(int index)
{
    return get_leg(index)->get_point_b();
}

double SuitLegList::get_start_time(int index)
{
    return get_leg(index)->get_start_time();
}

int SuitLegList::get_leg_index_at_time(double time, int start_leg)
{
    if (start_leg < 0 || start_leg >= m_legs.size() || get_leg_time(start_leg) > time)
        start_leg = 0;

    int num_legs = get_num_legs();
    int next_leg = start_leg + 1;

    while (true)
    {
        if (next_leg >= num_legs || get_start_time(next_leg) > time)
            break;

        next_leg++;
    }

    return next_leg - 1;
}

bool SuitLegList::is_point_in_range(PT(DNASuitPoint) point, double low_time, double high_time)
{
    point_index_t point_index = point->get_index();
    int start_leg_index = get_leg_index_at_time(low_time, 0);
    int end_leg_index = get_leg_index_at_time(high_time, point_index);

    for (int i = start_leg_index; i <= end_leg_index; i++)
    {
        if (i >= get_num_legs())
            return false;

        SuitLeg* leg = get_leg(i);
        if (leg->get_point_a()->get_index() == point_index)
            return true;

        else if (leg->get_point_b()->get_index() == point_index)
            return true;
    }

    return false;
}

SuitLeg::LegType SuitLegList::get_next_leg_type(int index)
{
    return get_suit_leg_type(m_path->get_point(index)->get_point_type(),
                             m_path->get_point(index + 1)->get_point_type());
}

void SuitLegList::add_leg(PT(DNASuitPoint) point_a, PT(DNASuitPoint) point_b,
                          SuitLeg::LegType type, block_number_t landmark_building_index)
{
    zone_id_t zone_id = m_store->get_suit_edge_zone(point_a->get_index(),
                                                    point_b->get_index());

    SuitLeg* last_leg = m_legs.size() ? m_legs[m_legs.size() - 1] : nullptr;
    double start_time = last_leg ? last_leg->get_start_time() + last_leg->get_leg_time() : 0.0;

    if (point_a->get_point_type() == DNASuitPoint::COGHQ_OUT_POINT && type == SuitLeg::T_walk)
    {
        SuitLeg* hq_leg = new SuitLeg(start_time, zone_id,
                                      point_a->get_landmark_building_index(),
                                      point_a, point_b,
                                      SuitLeg::T_from_cog_HQ,
                                      m_suit_walk_speed);
        m_legs.push_back(hq_leg);
        start_time += hq_leg->get_leg_time();
    }

    SuitLeg* leg = new SuitLeg(start_time, zone_id,
                               landmark_building_index,
                               point_a, point_b, type,
                               m_suit_walk_speed);
    m_legs.push_back(leg);

    if (point_b->get_point_type() == DNASuitPoint::COGHQ_IN_POINT && type == SuitLeg::T_walk)
    {
        m_legs.push_back(new SuitLeg(start_time + leg->get_leg_time(), zone_id,
                                     point_b->get_landmark_building_index(),
                                     point_a, point_b,
                                     SuitLeg::T_to_cog_HQ,
                                     m_suit_walk_speed));
    }
}
